lib/action_state/predicate.rb

```diff
@@ -1,50 +1,41 @@
 module ActionState
   class Predicate
-    attr_reader :result
-
     def initialize(record)
       @record = record
-      @result = true
+      @match_refuted = false
     end
 
-    def excluding(*records)
-      return self unless result
-
-      @result = false if records.include?(@record)
+    def matched?
+      !@match_refuted
+    end
 
+    def excluding(*records)
+      @match_refuted ||= records.include?(@record)
       self
     end
 
     def where(**kwargs)
-      return self unless @result
-
-      @result = false if kwargs.any? { |k, v| !compare(v, @record.send(k)) }
-
+      @match_refuted ||= kwargs.any? { |k, v| !compare(v, @record.send(k)) }
```

Comment:
I don’t fully understand why we can’t use `none?` here, as that would make this read better and you could even pass the same proc to `any?` and `none?`. But I know when I tried this before a few tests failed. 🤔

Reply:
Yeah, same. I don't get it either.

Reply:
I went as far as looking at the Ruby implementation to see that it wasn’t just.

```ruby
def none?(&block)
  !any?(&block)
end
```

I should dig into this a bit more and see what’s going on. It’s super confusing.

Reply:
I found out why! It's because `none?` returns true for the empty collection too. Confirmed in irb:

```irb
>> [].none?
=> true
>> [].none? { true }
=> true # true is returned because the block is never executed since there's no elements in the Enumerable.
>> [1].none? { true }
=> false
```

The fix is:

```ruby
    def where(**kwargs)
      @match_refuted ||= kwargs.any? && kwargs.none? { |k, v| compare(v, @record.send(k)) }
      self
    end
```

Reply:
Thinking about this, it couldn’t possibly be `!any` without being really inefficient. The result of the block actually needs to be reversed. I would have thought it would be equivalent to

```ruby
def none?(&block)
  any? { !block.call }
end
```